src/passes/Precompute.cpp

```diff
@@ -203,26 +206,24 @@ struct Precompute
         // this expression causes a return. if it's already a return, reuse the
         // node
         if (auto* ret = curr->dynCast<Return>()) {
-          if (flow.getSingleValue().type != Type::none) {
+          if (flow.values.size() > 0) {
```

Comment:
can this line keep comparing to none? maybe we should add `isNone` alongside `isConcrete` etc.

one risk that avoids is that this diff as it stands makes us check the actual values as opposed to the declared type. a corner case there is if the flow is breaking, then the declared type might have values, but the actual values are empty.

(similar checks later down)

Reply:
It could, but calculating the type to compare to None is more expensive than just looking at the number of values.

>  this diff as it stands makes us check the actual values as opposed to the declared type.

I don't think this is correct. This diff just replaces or protects uses of `flow.getSingleValue()`, which already examines the values in the flow. Flows do not even store their expected types.

Reply:
What do you mean by "calculating the type"? Wouldn't an `isNone` (replacing `getSingleValue`) just immediately check if the type is none?

>  This diff just replaces or protects uses of flow.getSingleValue(), which already examines the values in the flow.

The inner if protects that way, yeah, but the outer one is what worries me. If the type is concrete but the flow is breaking, then in the past the outer if would get true, while after this patch it gets false.

But rereading the code i see this flow is breaking anyhow, so that looks ok.

Reply:
> Wouldn't an isNone (replacing getSingleValue) just immediately check if the type is none?

Do you mean implement `isNone` by checking whether there are zero values? That would work and be quick. I thought you were thinking of examining the Type of the Flow, though. Since Flows do not have Types, a Type would have to be created by collecting the Types of each value in the Flow. That would be a more expensive operation, since it at the very least requires a vector allocation and a hashmap lookup for the memoized type.

Reply:
Oh, thanks, that's what I was missing - that Flows don't track their types. Which makes sense.